### PR TITLE
[Wait for #1381] [Dataset] Introduce data producer

### PR DESCRIPTION
**Pull request by zhoonit**
**[Pending commit: #1381]**

- [Dataset] Introduce data producer

```
This patch introduces data producer which abstracts a single iteration
creator.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: nntrainer/dataset/data_producers.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    std::numeric_limits<unsigned long long>::max();
+
+  /**
+   * @brief Destroy the Data Loader object
+   *
+   */
+  virtual ~DataProducer() {}
+
+  /**
+   * @brief Get the layer type
```
Comment (jijoongmoon) [inline nntrainer/dataset/data_producers.h:53]: layer type?

**Review by jijoongmoon — APPROVED** [review]
LGTM with mior comment.

File: nntrainer/dataset/data_producers.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace nntrainer {
+
+/**
+ * @brief DataProducer interface used to abstract data provider
+ *
+ */
+class DataProducer {
+public:
+  using Iteration = std::tuple<bool, std::vector<Tensor>, std::vector<Tensor>>;
```
Comment (kparichay) [inline nntrainer/dataset/data_producers.h:33]: please add a comment describing/naming the three type variables.

File: nntrainer/dataset/data_producers.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+   * @return const std::string type representation
+   */
+  virtual const std::string getType() const = 0;
+
+  /**
+   * @brief Set the Property object
+   *
+   * @param properties properties to set
+   */
+  virtual void setProperty(const std::vector<std::string> &properties) {}
```
Comment (kparichay) [inline nntrainer/dataset/data_producers.h:63]: how about making `setProperty()` either pure virtual or make the default implementation simply throw?

File: nntrainer/dataset/data_producers.h [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+  /**
+   * @brief create an iteration
+   * @return Iteration iteration, if std::get<0>(retval) == true means end of
+   * iteration, at the end of the iteration, it's responsibility of @a this to
+   * shuffle.
+   */
+  using Gernerator = std::function<Iteration(void)>;
+
+  constexpr inline static unsigned long long SIZE_UNDEFINED =
```
Comment (kparichay) [inline nntrainer/dataset/data_producers.h:43]: how about `size_t`?

File: nntrainer/dataset/data_producers.h [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+   * cannot be determined, this function must return
+   * DataProducer::SIZE_UNDEFINED;
+   *
+   * @param input_dims input dimensions
+   * @param label_dims label dimensions
+   *
+   * @return size calculated size
+   */
+  virtual unsigned long long
+  size(const std::vector<TensorDim> &input_dims,
```
Comment (kparichay) [inline nntrainer/dataset/data_producers.h:85]: related comment - https://github.com/nnstreamer/nntrainer/pull/1391/commits/2326033c83838d910d8160797df4c0874e7e3552#r671962797

File: nntrainer/dataset/data_producers.h (hunk 2 above)
Comment (zhoonit) [reply nntrainer/dataset/data_producers.h:33]: Done

File: nntrainer/dataset/data_producers.h (hunk 1 above)
Comment (zhoonit) [reply nntrainer/dataset/data_producers.h:53]: Changed :)

File: nntrainer/dataset/data_producers.h (hunk 3 above)
Comment (zhoonit) [reply nntrainer/dataset/data_producers.h:63]: I'll make this throw if property is not empty, this is similar to noop for data producers that are not using setProperty simply it does not neeed it.

File: nntrainer/dataset/data_producers.h (hunk 4 above)
Comment (zhoonit) [reply nntrainer/dataset/data_producers.h:43]: any reasons? I chose to use unsigned long long as the value is predictable but size_t is not.

File: nntrainer/dataset/data_producers.h (hunk 5 above)
Comment (zhoonit) [reply nntrainer/dataset/data_producers.h:85]: Sorry, I can't see the comment, if this is in the next PR, I'll automatically check though :)

File: nntrainer/dataset/data_producers.h (hunk 4 above)
Comment (kparichay) [reply nntrainer/dataset/data_producers.h:52]: size_t is standard to store the size of the dataset and scales its byte size appropriately as per systems. For example, most the STL data structures in C++ use size_t to give out size.

File: nntrainer/dataset/data_producers.h (hunk 5 above)
Comment (kparichay) [reply nntrainer/dataset/data_producers.h:98]: updated the link.

File: nntrainer/dataset/data_producers.h (hunk 3 above)
Comment (kparichay) [reply nntrainer/dataset/data_producers.h]: works.

**Review by kparichay — APPROVED** [review]
LGTM

**Pull request merged**